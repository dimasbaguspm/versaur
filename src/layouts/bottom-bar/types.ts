import type { HTMLAttributes } from 'react'

/**
 * Props for the BottomBar container
 */
export interface BottomBarProps extends HTMLAttributes<HTMLElement> {
  /**
   * Visual variant for the BottomBar
   * @default 'primary'
   */
  variant?: 'primary' | 'secondary' | 'ghost' | 'neutral' | 'tertiary'
  /**
   * Size of the BottomBar
   * @default 'md'
   */
  size?: 'sm' | 'md' | 'lg'
}

/**
 * Props for each BottomBar.Item
 */
export interface BottomBarItemProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Icon element for the item
   */
  icon: React.ReactNode
  /**
   * Label for the item
   */
  label?: string
  /**
   * Whether the item is active
   */
  active?: boolean
}
