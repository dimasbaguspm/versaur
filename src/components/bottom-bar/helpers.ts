import { cva } from '@/utils/variants'

export const bottomBarVariants = cva(
  'absolute bottom-0 left-0 w-full flex justify-between items-center bg-white border-t border-coral/10 z-40 safe-bottom px-6 py-1',
  {
    variants: {
      variant: {
        primary: '',
        secondary: '',
        ghost: '',
        neutral: '',
        tertiary: '',
      },
      size: {
        sm: 'h-12 text-sm',
        md: 'h-16 text-base',
        lg: 'h-20 text-lg',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

export const bottomBarItemVariants = cva(
  'flex flex-col items-center justify-center gap-1 px-2 py-1 rounded transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-primary',
  {
    variants: {
      active: {
        true: 'font-semibold text-coral',
        false: 'text-slate',
      },
    },
    defaultVariants: {
      active: false,
    },
  }
)
