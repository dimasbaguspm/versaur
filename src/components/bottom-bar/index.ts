export { BottomBar } from './bottom-bar'
export type { BottomBarProps, BottomBarItemProps } from './types'
