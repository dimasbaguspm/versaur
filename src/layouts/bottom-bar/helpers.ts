import { cva } from '@/utils/variants'

export const bottomBarVariants = cva(
  'sticky bottom-0 left-0 w-full flex justify-between items-center bg-white border-t border-border shadow-xs z-40 safe-bottom px-8 py-1',
  {
    variants: {
      variant: {
        primary: '',
        secondary: '',
        ghost: '',
        neutral: '',
        tertiary: '',
      },
      size: {
        sm: 'h-12 text-sm',
        md: 'h-16 text-base',
        lg: 'h-20 text-lg',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

export const bottomBarItemVariants = cva(
  'flex flex-col items-center justify-center gap-1 px-2 py-1',
  {
    variants: {
      active: {
        true: 'font-semibold text-primary',
        false: 'text-ghost',
      },
    },
    defaultVariants: {
      active: false,
    },
  }
)
