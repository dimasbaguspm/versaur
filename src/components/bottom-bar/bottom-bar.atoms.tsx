/**
 * BottomBar.Item sub-component
 */
import { forwardRef } from 'react'
import type { BottomBarItemProps } from './types'
import { bottomBarItemVariants } from './helpers'

export const BottomBarItem = forwardRef<HTMLButtonElement, BottomBarItemProps>(
  ({ icon, label, active = false, className, ...props }, ref) => (
    <button
      ref={ref}
      type='button'
      aria-current={active ? 'page' : undefined}
      className={bottomBarItemVariants({ active, className })}
      {...props}
    >
      {icon}
      {label && <span className='text-xs mt-0.5'>{label}</span>}
    </button>
  )
)
