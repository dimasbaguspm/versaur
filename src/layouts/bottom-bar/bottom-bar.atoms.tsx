/**
 * BottomBar.Item sub-component
 */
import { forwardRef } from 'react'
import type { BottomBarItemProps } from './types'
import { bottomBarItemVariants } from './helpers'

export const BottomBarItem = forwardRef<HTMLButtonElement, BottomBarItemProps>(
  (
    {
      as: Component = 'button',
      icon,
      label,
      active = false,
      className,
      children,
      ...props
    },
    ref
  ) => (
    <Component
      // @ts-expect-error it's parameterized component
      ref={ref}
      aria-current={active ? 'page' : undefined}
      className={bottomBarItemVariants({ active, className, as: Component })}
      {...props}
    >
      {icon && icon}
      {children && children}
      {label && <span className='text-xs mt-0.5'>{label}</span>}
    </Component>
  )
)
