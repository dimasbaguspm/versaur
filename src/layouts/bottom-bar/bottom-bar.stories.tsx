/**
 * Storybook stories for BottomBar (Layout group)
 *
 * - Demonstrates BottomBar with 3 and 5 items
 * - Follows Versaur design system and accessibility standards
 * - Icons use lucide-react (dev only)
 *
 * @group Layout/BottomBar
 */
import type { Meta, StoryObj } from '@storybook/react'
import { BottomBar } from './bottom-bar'
import {
  PlusIcon,
  HomeIcon,
  SearchIcon,
  SettingsIcon,
  UserIcon,
} from 'lucide-react'
import { ButtonIcon } from '@/primitive/button-icon'
import { Avatar, Icon } from '@/primitive'
import { useState } from 'react'

const meta: Meta<typeof BottomBar> = {
  title: 'Layouts/BottomBar',
  component: BottomBar,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component:
          'BottomBar is a mobile-first navigation bar for layout grouping, following Material guidelines and Versaur design system.',
      },
    },
  },
  decorators: [
    Story => (
      <div
        style={{
          minHeight: '400px',
          background: 'white',
          margin: '1rem',
          border: '1px var(--color-primary) solid',
          position: 'relative',
          display: 'flex',
          flexDirection: 'column',
        }}
      >
        <div style={{ flex: 1 }} />
        <Story />
      </div>
    ),
  ],
}
export default meta

export const ThreeItems: StoryObj<typeof BottomBar> = {
  name: '3 Items',
  render: function ThreeItems() {
    const [activeItem, setActiveItem] = useState(0)

    return (
      <BottomBar>
        <BottomBar.Item
          icon={<Icon as={HomeIcon} color='inherit' />}
          active={activeItem === 0}
          onClick={() => setActiveItem(0)}
        />
        <BottomBar.Item
          icon={<Icon as={SearchIcon} color='inherit' />}
          active={activeItem === 1}
          onClick={() => setActiveItem(1)}
        />
        <BottomBar.Item
          icon={<Icon as={UserIcon} color='inherit' />}
          active={activeItem === 2}
          onClick={() => setActiveItem(2)}
        />
      </BottomBar>
    )
  },
}

export const FiveItems: StoryObj<typeof BottomBar> = {
  name: '5 Items',
  render: function Component() {
    const [activeItem, setActiveItem] = useState(0)

    return (
      <BottomBar>
        <BottomBar.Item
          icon={<Icon as={HomeIcon} color='inherit' />}
          active={activeItem === 0}
          onClick={() => setActiveItem(0)}
        />
        <BottomBar.Item
          icon={<Icon as={SearchIcon} color='inherit' />}
          active={activeItem === 1}
          onClick={() => setActiveItem(1)}
        />
        <BottomBar.Item
          as='div'
          icon={
            <ButtonIcon
              as={PlusIcon}
              variant='primary'
              shape='circle'
              aria-label='Add Transaction'
              onClick={() => setActiveItem(2)}
            />
          }
        />
        <BottomBar.Item
          icon={<Icon as={SettingsIcon} color='inherit' />}
          active={activeItem === 3}
          onClick={() => setActiveItem(3)}
        />
        <BottomBar.Item
          active={activeItem === 4}
          onClick={() => setActiveItem(4)}
        >
          <Avatar size='sm'>DM</Avatar>
        </BottomBar.Item>
      </BottomBar>
    )
  },
}
