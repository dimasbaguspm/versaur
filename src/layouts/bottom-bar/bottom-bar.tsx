import type { BottomBarProps } from './types'
import { bottomBarVariants } from './helpers'
import { BottomBarItem } from './bottom-bar.atoms'

const BottomBarRoot = ({
  children,
  variant = 'primary',
  size = 'md',
  className,
  ...props
}: BottomBarProps) => (
  <nav
    role='navigation'
    aria-label='Bottom navigation'
    className={bottomBarVariants({ variant, size, className })}
    {...props}
  >
    {children}
  </nav>
)

export const BottomBar = Object.assign(BottomBarRoot, {
  Item: BottomBarItem,
})
