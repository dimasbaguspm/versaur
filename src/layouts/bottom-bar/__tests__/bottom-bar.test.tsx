/**
 * BottomBar component tests
 *
 * - Uses composeStories from @storybook/react
 * - Asserts accessibility and snapshot
 * - Follows Versaur UI guidelines
 */
import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../bottom-bar.stories'

const { ThreeItems, FiveItems } = composeStories(stories)

describe('BottomBar', () => {
  it('renders 3 items and matches snapshot', () => {
    const { asFragment } = render(<ThreeItems />)
    expect(screen.getByText('Home')).toBeInTheDocument()
    expect(screen.getByText('Search')).toBeInTheDocument()
    expect(screen.getByText('Profile')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders 5 items and matches snapshot', () => {
    const { asFragment } = render(<FiveItems />)
    expect(screen.getByText('Home')).toBeInTheDocument()
    expect(screen.getByText('Search')).toBeInTheDocument()
    expect(screen.getByText('Settings')).toBeInTheDocument()
    expect(screen.getByText('Profile')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('has accessible role navigation', () => {
    render(<ThreeItems />)
    expect(screen.getByRole('navigation')).toBeInTheDocument()
  })
})
