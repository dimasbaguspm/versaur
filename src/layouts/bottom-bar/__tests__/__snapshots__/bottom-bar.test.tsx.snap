// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomBar > renders 3 items and matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="min-height: 400px; background: white;"
  >
    <nav
      aria-label="Bottom navigation"
      class="sticky bottom-0 left-0 w-full flex justify-between items-center bg-white border-t border-border shadow-xs z-40 safe-bottom px-8 py-1 h-16 text-base"
      role="navigation"
    >
      <div
        aria-current="page"
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 font-semibold text-coral"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-house"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
          />
          <path
            d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Home
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Search
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-user"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Profile
        </span>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;

exports[`BottomBar > renders 5 items and matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="min-height: 400px; background: white;"
  >
    <nav
      aria-label="Bottom navigation"
      class="sticky bottom-0 left-0 w-full flex justify-between items-center bg-white border-t border-border shadow-xs z-40 safe-bottom px-8 py-1 h-16 text-base"
      role="navigation"
    >
      <div
        aria-current="page"
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 font-semibold text-coral"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-house"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
          />
          <path
            d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Home
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Search
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <button
          aria-disabled="false"
          aria-label="Add Transaction"
          class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 w-10 rounded-full"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-plus inline-flex items-center justify-center w-5 h-5 text-neutral"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-settings"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Settings
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 px-2 py-1 text-slate"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-user"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
        <span
          class="text-xs mt-0.5"
        >
          Profile
        </span>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
